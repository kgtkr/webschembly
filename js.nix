{ lib, ... }: {
  perSystem = { self', pkgs, ... }:
    let
      nodejs = pkgs.nodejs_22;
      src = lib.fileset.toSource {
        root = ./.;
        fileset = lib.fileset.unions ([
          ./package.json
          ./package-lock.json
        ] ++ lib.map (lib.path.append ./.) (builtins.fromJSON (builtins.readFile ./package.json)).workspaces);
      };
      generatedSrc = pkgs.napalm.buildPackage src {
        inherit nodejs;
        name = "webschembly-node_modules";
      };
      webschembly-playground = pkgs.stdenv.mkDerivation {
        name = "webschembly-playground";
        buildInputs = [ pkgs.gnumake nodejs ];
        src = "${generatedSrc}/_napalm-install";
        buildPhase = ''
          make -C webschembly-playground WEBSCHEMBLY_RUNTIME=${self'.packages.webschembly-runtime}/lib/webschembly_runtime.wasm
        '';
        installPhase = ''
          mkdir -p $out
          cp -r webschembly-playground/dist/* $out
        '';
      };
    in
    {
      packages = {
        inherit webschembly-playground;
        webschembly-playground-for-pages = webschembly-playground.overrideAttrs (oldAttrs: {
          BASE_URL = "/webschembly/";
        });
      };
      make-shells.default = {
        packages = [
          nodejs
        ];
      };
    };
}
