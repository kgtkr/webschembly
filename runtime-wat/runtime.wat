(module
  (import "runtime" "init" (func $init))
  ;; TODO: 一旦大きめに確保、今後growする機能を追加
  (table (export "funcs") 10000 funcref)
  (start $init)
)
