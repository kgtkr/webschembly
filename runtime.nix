{ stdenv, binaryen, runtime-rust, runtime-wat }:
stdenv.mkDerivation {
  name = "runtime";
  dontUnpack = true;
  buildInputs = [ binaryen runtime-rust runtime-wat ];
  buildPhase = ''
    wasm-merge -o runtime.wasm ${runtime-rust}/lib/webschembly_runtime.wasm runtime ${runtime-wat}/lib/runtime.wasm runtime --enable-reference-types --enable-multimemory
  '';
  installPhase = ''
    mkdir -p $out/lib
    cp runtime.wasm $out/lib
  '';
}
