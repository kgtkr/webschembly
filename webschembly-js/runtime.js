import * as fs from "fs";
import * as path from "path";

export function createRuntime({
  runtimeName = "untitled",
  exit = process.exit,
  exitWhenException = true,
  logDir = process.env.LOG_DIR || null,
  runtimeBuf = fs.readFileSync(process.env["WEBSCHEMBLY_RUNTIME"]),
  writeBuf = (fd, buf) => {
    switch (fd) {
      case 1:
        process.stdout.write(buf);
        break;
      case 2:
        process.stderr.write(buf);
        break;
      default:
        throw new Error(`Unsupported file descriptor: ${fd}`);
    }
  },
  printEvalResult = false,
}) {
  const logBasename = path.basename(runtimeName) + "-" + Date.now();
  let logFile = null;
  if (logDir !== null) {
    try {
      fs.mkdirSync(logDir);
    } catch (e) {
      if (e.code !== "EEXIST") {
        throw e;
      }
    }

    logFile = fs.openSync(path.join(logDir, logBasename + ".log"), "a");
  }
  let instantiateCount = 0;

  const runtimeImportObjects = {
    js_instantiate: (bufPtr, bufSize) => {
      const buf = new Uint8Array(
        runtimeInstance.exports.memory.buffer,
        bufPtr,
        bufSize
      );
      if (logDir !== null) {
        fs.writeFileSync(
          path.join(logDir, logBasename + "-" + instantiateCount + ".wasm"),
          buf
        );
      }
      instantiateCount++;

      const instance = new WebAssembly.Instance(
        new WebAssembly.Module(buf),
        importObject
      );

      const result = instance.exports.start();
      if (printEvalResult) {
        runtimeInstance.exports.print_for_repl(result);
      }
    },
    js_webschembly_log: (bufPtr, bufLen) => {
      const s = new TextDecoder().decode(
        new Uint8Array(runtimeInstance.exports.memory.buffer, bufPtr, bufLen)
      );
      if (logFile !== null) {
        fs.writeSync(logFile, s + "\n");
      }
    },
    js_write_buf: (fd, bufPtr, bufLen) => {
      const buf = new Uint8Array(
        runtimeInstance.exports.memory.buffer,
        bufPtr,
        bufLen
      );
      writeBuf(1, buf);
    },
  };

  const runtimeInstance = new WebAssembly.Instance(
    new WebAssembly.Module(new Uint8Array(runtimeBuf)),
    {
      env: runtimeImportObjects,
    }
  );

  const importObject = {
    runtime: runtimeInstance.exports,
  };

  const errorHandle =
    (f) =>
    (...args) => {
      try {
        f(...args);
      } catch (e) {
        if (e instanceof WebAssembly.Exception) {
          if (e.is(runtimeInstance.exports.WEBSCHEMBLY_EXCEPTION)) {
            if (exitWhenException) {
              exit(1);
            }
          } else {
            throw e;
          }
        } else {
          throw e;
        }
      }
    };

  return {
    loadStdlib: errorHandle(() => {
      runtimeInstance.exports.load_stdlib();
    }),
    loadSrc: errorHandle((srcBuf) => {
      const srcBufPtr = runtimeInstance.exports.malloc(srcBuf.length);
      new Uint8Array(runtimeInstance.exports.memory.buffer).set(
        srcBuf,
        srcBufPtr
      );
      runtimeInstance.exports.load_src(srcBufPtr, srcBuf.length);
      runtimeInstance.exports.free(srcBufPtr);
    }),
    flushAll: () => {
      runtimeInstance.exports.flush_all();
    },
    cleanup: () => {
      runtimeInstance.exports.cleanup();
    },
  };
}
