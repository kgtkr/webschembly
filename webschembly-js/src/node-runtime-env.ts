import * as fs from "fs";
import * as path from "path";

export function createLogger({
  logDir = process.env.LOG_DIR || null,
  runtimeName = "untitled",
}) {
  const logBasename = path.basename(runtimeName) + "-" + Date.now();
  let logFile = null;
  if (logDir !== null) {
    try {
      fs.mkdirSync(logDir);
    } catch (e) {
      if (e.code !== "EEXIST") {
        throw e;
      }
    }

    logFile = fs.openSync(path.join(logDir, logBasename + ".log"), "a");
  }
  let instantiateCount = 0;

  return {
    instantiate: (buf) => {
      if (logDir !== null) {
        fs.writeFileSync(
          path.join(logDir, logBasename + "-" + instantiateCount + ".wasm"),
          buf
        );
      }
      instantiateCount++;
    },
    log: (s) => {
      if (logFile !== null) {
        fs.writeSync(logFile, s + "\n");
      }
    },
  };
}

export function createNodeRuntimeEnv({
  runtimeName = "untitled",
  exit = process.exit,
  logger = createLogger({ runtimeName }),
  runtimeModule = new WebAssembly.Module(
    fs.readFileSync(process.env["WEBSCHEMBLY_RUNTIME"])
  ),
  writeBuf = (fd, buf) => {
    switch (fd) {
      case 1:
        process.stdout.write(buf);
        break;
      case 2:
        process.stderr.write(buf);
        break;
      default:
        throw new Error(`Unsupported file descriptor: ${fd}`);
    }
  },
}) {
  return {
    runtimeName,
    exit,
    logger,
    runtimeModule,
    writeBuf,
  };
}
