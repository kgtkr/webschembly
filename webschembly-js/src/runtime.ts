export type RuntimeEnv = {
  exit: (code: number) => void;
  logger: RuntimeLogger;
  loadRuntimeModule: () => Promise<WebAssembly.Module>;
  writeBuf: (fd: number, buf: Uint8Array) => void;
};

export type RuntimeConfig = {
  exitWhenException?: boolean;
  printEvalResult?: boolean;
  compilerConfig?: CompilerConfig;
};

export type CompilerConfig = {
  enableJit?: boolean;
  enableJitOptimization?: boolean;
};

export function compilerConfigToString(config: CompilerConfig): string {
  return Object.entries(config)
    .filter(([, v]) => v !== undefined)
    .map(([k, v]) => `${k}=${v}`)
    .join(",");
}

export type RuntimeLogger = {
  log: (s: string) => void;
  instantiate: (buf: Uint8Array, ir: string | null) => void;
};

export type Runtime = {
  loadStdlib: () => void;
  loadSrc: (srcBuf: Uint8Array) => void;
  flushAll: () => void;
  cleanup: () => void;
  mallocString: (s: string) => [ptr: number, len: number];
  instance: TypedWebAssemblyInstance<RuntimeExports>;
};

export type RuntimeImportsEnv = {
  js_instantiate: (
    bufPtr: number,
    bufSize: number,
    irBufPtr: number,
    irBufSize: number,
    fromSrc: number
  ) => void;
  js_webschembly_log: (bufPtr: number, bufLen: number) => void;
  js_write_buf: (fd: number, bufPtr: number, bufLen: number) => void;
};

export type RuntimeImports = {
  env: RuntimeImportsEnv;
};

export type SchemeValue = { __tagSchemeValueBrand: any };

export type RuntimeExports = {
  memory: WebAssembly.Memory;
  WEBSCHEMBLY_EXCEPTION: WebAssembly.ExceptionTag;
  get_global: (namePtr: number, nameLen: number) => SchemeValue;
  new_args: (elemSize: number) => SchemeValue;
  set_args: (args: SchemeValue, index: number, value: number) => void;
  call_closure: (closure: SchemeValue, args: SchemeValue) => SchemeValue;
  malloc: (size: number) => number;
  free: (ptr: number) => void;
  load_stdlib: () => void;
  load_src: (srcPtr: number, srcLen: number) => void;
  flush_all: () => void;
  cleanup: () => void;
  init: () => void;
  compiler_config_enable_jit: (enable: number) => void;
  compiler_config_enable_jit_optimization: (enable: number) => void;
};

export type ModuleImports = {
  runtime: RuntimeExports;
  dynamic: Record<string, WebAssembly.ExportValue>;
};

export type ModuleExports = {
  start: () => number;
};

export type TypedWebAssemblyInstance<Exports> = WebAssembly.Instance & {
  exports: Exports;
};

export async function createRuntime(
  { exit, logger, loadRuntimeModule, writeBuf }: RuntimeEnv,
  {
    exitWhenException = true,
    printEvalResult = false,
    compilerConfig,
  }: RuntimeConfig
): Promise<Runtime> {
  const dynamic: Record<string, WebAssembly.ExportValue> = {};
  const runtimeImportObjects: RuntimeImportsEnv = {
    js_instantiate: (bufPtr, bufSize, irBufPtr, irBufSize, fromSrc) => {
      const buf = new Uint8Array(
        runtimeInstance.exports.memory.buffer,
        bufPtr,
        bufSize
      );
      const ir =
        irBufPtr === 0
          ? null
          : new TextDecoder().decode(
              new Uint8Array(
                runtimeInstance.exports.memory.buffer,
                irBufPtr,
                irBufSize
              )
            );

      logger.instantiate(buf, ir);

      const instance = new WebAssembly.Instance(
        new WebAssembly.Module(buf),
        importObject
      ) as TypedWebAssemblyInstance<ModuleExports>;
      Object.assign(dynamic, instance.exports);

      const result = instance.exports.start();
      if (printEvalResult && fromSrc !== 0) {
        const writeClosure = runtimeInstance.exports.get_global(
          writePtr,
          writeLen
        );
        const writeParams = runtimeInstance.exports.new_args(1);
        runtimeInstance.exports.set_args(writeParams, 0, result);
        runtimeInstance.exports.call_closure(writeClosure, writeParams);

        const newlineClosure = runtimeInstance.exports.get_global(
          newlinePtr,
          newlineLen
        );
        const newlineParams = runtimeInstance.exports.new_args(0);
        runtimeInstance.exports.call_closure(newlineClosure, newlineParams);
      }
    },
    js_webschembly_log: (bufPtr, bufLen) => {
      const s = new TextDecoder().decode(
        new Uint8Array(runtimeInstance.exports.memory.buffer, bufPtr, bufLen)
      );
      logger.log(s);
    },
    js_write_buf: (fd, bufPtr, bufLen) => {
      const buf = new Uint8Array(
        runtimeInstance.exports.memory.buffer,
        bufPtr,
        bufLen
      );
      writeBuf(fd, buf);
    },
  };

  const runtimeInstance = new WebAssembly.Instance(await loadRuntimeModule(), {
    env: runtimeImportObjects,
  } satisfies RuntimeImports) as TypedWebAssemblyInstance<RuntimeExports>;

  if (compilerConfig?.enableJit !== undefined) {
    runtimeInstance.exports.compiler_config_enable_jit(
      Number(compilerConfig.enableJit)
    );
  }
  if (compilerConfig?.enableJitOptimization !== undefined) {
    runtimeInstance.exports.compiler_config_enable_jit_optimization(
      Number(compilerConfig.enableJitOptimization)
    );
  }

  runtimeInstance.exports.init();

  const importObject: ModuleImports = {
    runtime: runtimeInstance.exports,
    dynamic,
  };

  const errorHandle =
    <A extends any[], R>(f: (...args: A) => R) =>
    (...args: A): void => {
      try {
        f(...args);
      } catch (e) {
        if (
          e instanceof WebAssembly.Exception &&
          e.is(runtimeInstance.exports.WEBSCHEMBLY_EXCEPTION)
        ) {
          if (exitWhenException) {
            exit(1);
          }
        } else {
          throw e;
        }
      }
    };

  function mallocString(s: string): [number, number] {
    const buf = new TextEncoder().encode(s);
    const bufPtr = runtimeInstance.exports.malloc(buf.length);
    new Uint8Array(runtimeInstance.exports.memory.buffer).set(buf, bufPtr);
    return [bufPtr, buf.length];
  }

  const [writePtr, writeLen] = mallocString("write");
  const [newlinePtr, newlineLen] = mallocString("newline");

  return {
    loadStdlib: errorHandle(() => {
      runtimeInstance.exports.load_stdlib();
    }),
    loadSrc: errorHandle((srcBuf) => {
      const srcBufPtr = runtimeInstance.exports.malloc(srcBuf.length);
      new Uint8Array(runtimeInstance.exports.memory.buffer).set(
        srcBuf,
        srcBufPtr
      );
      runtimeInstance.exports.load_src(srcBufPtr, srcBuf.length);
      runtimeInstance.exports.free(srcBufPtr);
    }),
    flushAll: () => {
      runtimeInstance.exports.flush_all();
    },
    cleanup: () => {
      runtimeInstance.exports.cleanup();
    },
    mallocString,
    instance: runtimeInstance,
  };
}
