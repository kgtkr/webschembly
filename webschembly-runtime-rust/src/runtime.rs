#[link(wasm_import_module = "runtime")]
extern "C" {
    pub fn throw_webassembly_exception();
}
