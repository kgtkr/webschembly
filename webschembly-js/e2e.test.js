import * as fs from "fs/promises";
import * as fsLegacy from "fs";
import { beforeAll, describe, expect, test } from "vitest";
import * as path from "path";
import { createRuntime } from "./runtime.js";
import { createNodeRuntimeEnv } from "./node-runtime-env.js";

function concatBufs(bufs) {
  const bufLen = bufs.map((buf) => buf.length).reduce((a, b) => a + b, 0);
  const resultBuf = new Uint8Array(bufLen);
  let offset = 0;
  for (const buf of bufs) {
    resultBuf.set(buf, offset);
    offset += buf.length;
  }
  return resultBuf;
}

const sourceDir = "src";
const filenames = fsLegacy
  .readdirSync(sourceDir)
  .filter((file) => file.endsWith(".scm"));

describe("E2E test", () => {
  let runtimeBuf;
  beforeAll(async () => {
    runtimeBuf = await fs.readFile(process.env["WEBSCHEMBLY_RUNTIME"]);
  });

  describe.each(filenames)("%s", (filename) => {
    let srcBuf;
    beforeAll(async () => {
      srcBuf = await fs.readFile(path.join(sourceDir, filename));
    });

    test("snapshot test", async () => {
      let exitCode = 0;
      const stdoutBufs = [];
      const stderrBufs = [];
      const runtime = createRuntime(
        createNodeRuntimeEnv({
          runtimeName: filename,
          exit: (code) => {
            exitCode = code;
          },
          writeBuf: (fd, buf) => {
            switch (fd) {
              case 1:
                stdoutBufs.push(new Uint8Array(buf));
                break;
              case 2:
                stderrBufs.push(new Uint8Array(buf));
                break;
              default:
                throw new Error(`Unsupported file descriptor: ${fd}`);
            }
          },
          runtimeBuf,
        }),
        {}
      );

      runtime.loadStdlib();
      runtime.loadSrc(srcBuf);
      runtime.cleanup();

      const stdout = new TextDecoder().decode(concatBufs(stdoutBufs));
      const stderr = new TextDecoder().decode(concatBufs(stderrBufs));

      expect({ exitCode, stdout, stderr }).toMatchSnapshot();
    });
  });
});
