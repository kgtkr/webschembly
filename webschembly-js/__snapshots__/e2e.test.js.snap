// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`E2E test > add.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "3
",
}
`;

exports[`E2E test > builtin_some_name.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "30
3
",
}
`;

exports[`E2E test > closure.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "42
45
",
}
`;

exports[`E2E test > define.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": ""Hello, World!"
",
}
`;

exports[`E2E test > define_builtin.scm > snapshot test 1`] = `
{
  "exitCode": 1,
  "stderr": "set! builtin is not allowed
",
  "stdout": "",
}
`;

exports[`E2E test > display.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "(2 . (() . ("hello" . ((1 . (2 . ())) . (abc . (efg . (abc . ())))))))
",
}
`;

exports[`E2E test > get_builtin.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "3
",
}
`;

exports[`E2E test > global_duplication.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "2
2
3
3
4
4
",
}
`;

exports[`E2E test > if.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": ""True1"
"False2"
",
}
`;

exports[`E2E test > lambda.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "42
",
}
`;

exports[`E2E test > list_syntax.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "()
(1 . (2 . (3 . (4 . ()))))
(1 . (2 . (3 . (4 . 5))))
(1 . (2 . (3 . (4 . ()))))
1
",
}
`;

exports[`E2E test > local_duplication.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "1
2
3
2
3
4
",
}
`;

exports[`E2E test > local_duplication2.scm > snapshot test 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "2
3
",
}
`;

exports[`E2E test > set_builtin.scm > snapshot test 1`] = `
{
  "exitCode": 1,
  "stderr": "set! builtin is not allowed
",
  "stdout": "",
}
`;
